src/halmos/__main__.py

```diff
@@ -761,21 +1021,73 @@ def run_contract(ctx: ContractContext) -> list[TestResult]:
 
         return []
 
+    # separate regular and invariant tests
+    test_funsigs, inv_funsigs = [], []
+    for sig in ctx.funsigs:
+        (inv_funsigs if sig.startswith("invariant_") else test_funsigs).append(sig)
+
+    test_results = []
+
+    # execute regular tests
+    if test_funsigs:
+        test_results.extend(run_tests(ctx, setup_ex, test_funsigs))
+
+    # execute invariant tests
+    if inv_funsigs:
+        test_results.extend(run_invariant_tests(ctx, setup_ex, inv_funsigs))
+
+    # reset any remaining solver states from the default context
+    setup_solver.reset()
+
+    return test_results
+
+
+def run_tests(
+    ctx: ContractContext,
+    pre_ex: Exec,
+    funsigs: list[str],
+    depth: int = 0,
+    terminal: bool = True,
+) -> list[TestResult]:
+    """
+    Executes each of the given test functions on the given input state.
+    Used for both regular and invariant tests.
+
+    Args:
+        ctx: The context of the test contract.
+        pre_ex: The input state from which each test will be run.
+        funsigs: A list of test function signatures to execute.
+        depth (optional, only for invariant testing): The current depth of the invariant testing run.
+        terminal (optional, only for invariant testing): A flag indicating whether this testing run is final.
+
+    Returns:
+        A list of test results.
+    """
+    args = ctx.args
+
     test_results = []
-    for funsig in ctx.funsigs:
+    for funsig in funsigs:
         selector = ctx.method_identifiers[funsig]
         fun_info = FunctionInfo(funsig.split("(")[0], funsig, selector)
         try:
             test_config = with_devdoc(args, funsig, ctx.contract_json)
+            # TODO: reuse solver across different functions
             solver = mk_solver(test_config)
             debug(f"{test_config.formatted_layers()}")
```

Comment:
we probably need to remove this during an invariant run, or maybe have a separate `--debug-config` flag?

Reply:
i agree. i prefer a separate flag for this, or having a multi-level debug flag for longer term.